packages/grid/x-grid/src/tests/selection.XGrid.test.tsx

```diff
@@ -126,4 +133,30 @@ describe('<XGrid /> - Selection', () => {
     });
     expect(apiRef.current.getSelectedRows()).to.have.keys([0]);
   });
+
+  it('should select only filtered rows after filter is applied', () => {
+    render(<Test checkboxSelection />);
+    const selectAll = screen.getByRole('checkbox', {
+      name: /select all rows checkbox/i,
+    });
+    apiRef.current.setFilterModel({
+      items: [
+        {
+          columnField: 'brand',
+          operatorValue: 'contains',
+          value: 'Puma',
+        },
+      ],
+    });
+    expect(getColumnValues(1)).to.deep.equal(['Puma']);
+    fireEvent.click(selectAll);
+    // TODO fix, should be only 2
```

Comment:
Yup, this gets fixed by removing `React.memo` after which we get the updated visible rows in `GridHeaderCheckbox`.